src/main/java/gyro/google/compute/FirewallRuleResource.java

```diff
@@ -0,0 +1,580 @@
+            if (!getDestinationRanges().isEmpty()) {
+                errors.add(new ValidationError(this, "destination-ranges", "'destination-ranges' cannot be set when 'direction' set to 'INGRESS'"));
+            }
+
+            if (!getTargetTags().isEmpty()) {
+                errors.add(new ValidationError(this, "target-tags", "'target-tags' cannot be set when 'direction' set to 'INGRESS'"));
+            }
+
+            if (!getTargetServiceAccounts().isEmpty()) {
+                errors.add(new ValidationError(this, "target-service-account", "'target-service-account' cannot be set when 'direction' set to 'INGRESS'"));
+            }
+
+        } else if (getDirection().equals("EGRESS")) {
+            if (!getSourceRanges().isEmpty()) {
+                errors.add(new ValidationError(this, "source-ranges", "'source-ranges' cannot be set when 'direction' set to 'EGRESS'"));
+            }
+
+            if (!getSourceTags().isEmpty()) {
+                errors.add(new ValidationError(this, "source-tags", "'source-tags' cannot be set when 'direction' set to 'EGRESS'"));
+            }
+
+            if (!getSourceServiceAccounts().isEmpty()) {
+                errors.add(new ValidationError(this, "source-service-account", "'source-service-account' cannot be set when 'direction' set to 'EGRESS'"));
+            }
+        }
+
+        if ((!getTargetTags().isEmpty() || !getSourceTags().isEmpty()) && !getTargetServiceAccounts().isEmpty()) {
+            errors.add(new ValidationError(this, "target-service-accounts", "'target-service-accounts' cannot be set when 'target-tags' or 'source-tags' are set"));
+        }
+
+        return errors;
+    }
+
+    /**
+     * Creates a Firewall object that is consumed by the create, update and patch api.
+     * Divides the construction of the firewall object based on actionType.
+     * Fields 'description', 'source ranges', 'source tags', 'target tags' and 'allowed' can only be updated using the update api.
+     * Fields 'destination ranges', 'disabled', 'priority', 'source service accounts', 'log config', 'target service account' and 'denied' can only be updated using the patch api.
+     *
+     * @param client The compute client
+     * @param actionType The type of action (CREATE, UPDATE or PATCH)
+     * @return Firewall object
+     */
+    private Firewall toFirewall(Compute client, ACTION_TYPE actionType) {
+        Firewall firewall = getFirewall(client, actionType);
+
+        if (actionType.equals(ACTION_TYPE.UPDATE) || actionType.equals(ACTION_TYPE.CREATE)) {
+            firewall.setDescription(getDescription());
+            firewall.setSourceRanges(!getSourceRanges().isEmpty() ? new ArrayList<>(getSourceRanges()) : null);
+            firewall.setSourceTags(!getSourceTags().isEmpty() ? new ArrayList<>(getSourceTags()) : null);
+            firewall.setTargetTags(!getTargetTags().isEmpty() ? new ArrayList<>(getTargetTags()) : null);
+
+            if (getRuleType().equals("ALLOW")) {
```

Comment:
I noticed that for `ACTION_TYPE.UPDATE` you process `ALLOW` and in the following block for `ACTION_TYPE.PATCH` you process `DENY`. 

It seems like something is wrong with the use of the APIs.

Reply:
As discussed above, the update API allows updating `allowed` while the patch API allows updating `denied`.

Reply:
We figured that `patch` is able to update all the fields and so the `gyro update` now uses patch to update all the fields for the firewall rule. So the above code is replaced